### ISSUE TITLE
Beta Distribution values wrong for a=b---> 0

### ISSUE THREAD

**Issue by trdavidson** [pytorch/pytorch#15738]
## To Reproduce

copy paste the following in a notebook:

```
import torch
from scipy.stats import beta as beta_sp
from torch.distributions import Beta as beta_t
from matplotlib import pyplot as plt

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
dtype = torch.float32

a = 1e-2
b = 1e-2
size = [10000]

x1 = beta_sp.rvs(a, b, size=size)
x2 = beta_t(torch.tensor(a, device=device, dtype=dtype), 
            torch.tensor(b, device=device, dtype=dtype)).sample(torch.Size(size))

def plot_compare(x1, x2):
    plt.hist([x1, x2.cpu().numpy()], label=['scipy', 'pytorch'], density=True, alpha=0.75);
    plt.legend(loc='best')
    plt.title('Beta distribution for values a=b=%.3f' % a)
    plt.show()
    
# wrong as a=b --> 0, starting at 1e-2
plot_compare(x1, x2)
```
(1) for `device = cpu`, there is an error peak around 0.55  
![cpu_beta](https://user-images.githubusercontent.com/7077018/50701764-17d7f300-104f-11e9-869c-a3700f3c11e5.png)
(2) for `dvice = cuda`, there is a bias towards 1.
![cuda_beta](https://user-images.githubusercontent.com/7077018/50701763-17d7f300-104f-11e9-8a69-002f0e87f945.png)

## Expected behavior

see the `scipy` output for expected values.

## Environment
 - PyTorch Version (e.g., 1.0): 1.0
 - OS (e.g., Linux): mac
 - How you installed PyTorch: `conda`
 - Python version: 3.6
 - CUDA/cuDNN version: 9.0
 - GPU models and configuration: TitanX

cc @vincentqb @fritzo @neerajprad @alicanb @vishwakftw

**Comment by ajaech** [comment]
I did some investigation to find out why there is a peak in the CPU implementation at 0.5. (Note that the histogram above is slightly misleading. The peak is at 0.5 not 0.55.)

The root cause appears to be underflow in the implementation of the sample_gamma function. The beta distribution is a special case of the dirichlet distribution. And dirichlet uses the gamma distribution to pick samples.

https://github.com/pytorch/pytorch/blob/de0784510d06504d0825112e003370070ecdcd7d/aten/src/ATen/native/Distributions.h#L89

**Comment by trdavidson** [comment]
@ajaech apologies for not picking my bins more carefully in the histogram, it's 0.5 indeed. Interesting that the `standard_gamma` is still not working properly, given the discussion and closed issue in https://github.com/pytorch/pytorch/issues/12373#issuecomment-427681753 .

Any clue why the `gpu` implementation would be biased?

**Comment by fritzo** [comment]
The reason for the peak in the center is that `Beta` avoids overflow by clamping very small values to be at least `torch.finfo(...).tiny`. In extreme cases of very small parameters, most values get clamped, whence the center peak `tiny / (tiny + tiny) = 0.5`. I agree with @alicanb that the easiest way to deal with these extremely small parameters is to use higher-precision via `torch.float64`.

@ajaech #12373 is an unrelated bug in expanded tensors that affected `Beta` as well as many other functions, whereas this issue is specifically poor sampling distribution due to underflow.

**Comment by ailzhang** [comment]
@fritzo Can you send a patch for this? Thanks!

**Comment by fritzo** [comment]
@ailzhang I'm not sure the best way to fix the CPU issue, and I still don't understand the GPU issue. I'll think about it though. Any suggestions?

**Comment by fmassa** [comment]
@fritzo if I understand it correctly, the kernels in `sample_gamma` already use `accscalar_t` for internal computations, so maybe one solution would be to perform the normalization + clamping in float64 in
https://github.com/pytorch/pytorch/blob/7c08f1083e56ce7e9799a15b36947c46052f0058/torch/distributions/dirichlet.py#L13-L14
and then convert back to the original tensor type? But this might be suboptimal

**Comment by fritzo** [comment]
@fmassa That sounds like a great solution. I think we could easily implement a new `torch._sample_dirichlet` that calls the [sample_gamma() from Distributions.h](https://github.com/pytorch/pytorch/blob/de07845/aten/src/ATen/native/Distributions.h#L89) by:
1. cast `alpha` to `accscalar_t`;
2. generate an `accscalar_t` sample from `standard_gamma`;
3. `clamp(min=accscalar_t.tiny)`;
4. normalize;
5. cast to `scalar_t`;
6. `clamp(min=scalar_t.tiny)`;

Finally we can replace `_dirichlet_sample_nograd` with `torch._sample_dirichlet` in [dirichlet.py](https://github.com/pytorch/pytorch/blob/de07845/torch/distributions/dirichlet.py#L11). Since `Beta` is implemented in terms of `Dirichlet`, the `Beta` issues should thus be resolved.

Is this something you can implement? I would need help implementing this (especially the C API bindings for `_sample_dirichlet`), since things have moved around so much since I implemented `standard_gamma` etc.

**Comment by alicanb** [comment]
This is fixed, right?

**Comment by cpuhrsch** [comment]
It looks like referenced PRs resolved this issue. @trdavidson - if this is not the case please feel encouraged to reopen this issue.

**Comment by kulalexey** [comment]
Get wrong distribution for a=b=1e-2 on cuda with float32.
For a=b=1e-3 bug exist on both cuda and cpu even with float64
pytorch: 1.4.0
cuda: 10.2

**Comment by alicanb** [comment]
I can reproduce this in master

**Comment by neerajprad** [comment]
The issue is the same as mentioned by @fritzo above and, while not as frequent, it exists whenever `_s_gamma_cpu` returns a tensor where all elements in the event dim are tiny (clamped to `2.2251e-308` for double). I'm not sure what's the best way to handle this.

**Comment by rasa200** [comment]
Hi!

I just want to comment that we run in a similar issue in Rust: small parameters for Beta leads to erroneous samplings. The fix was to change the sampling algorithm for something more robust. We are still improving the performance though. 

You can see the issue [here](https://github.com/rust-random/rand/issues/999) and the solution [here](https://github.com/rust-random/rand/pull/1000#issuecomment-666436210).

I am not sure if you have the same problem but hope this is somehow useful.

Btw, I ended up here searching for something I could run in the GPU.

**Comment by bytesnake** [comment]
For `pytorch==2.3.1` and f64 dtype the issue still persists ony my side. The numpyro library provides a stable version for $\alpha \to 0$ which is easy to port to pytorch
https://github.com/fehiepsi/numpyro/blob/04c8fcb12539439008949c17f233e28dd71310fe/numpyro/distributions/continuous.py#L141-L150

```python
import torch
from torch.distributions import Exponential, Gamma

def dirichlet_stable(concentration, shape):
    exp = Exponential(torch.tensor([1.0])).rsample(shape + concentration.shape)[..., 0].to(concentration.device)
    gamma = Gamma(1 + concentration, rate=1.0).rsample(shape)

    tmp = torch.softmax(gamma.log() - exp / concentration, dim=-1)
    finfo = torch.finfo(concentration.dtype)
    return tmp.clamp(min=finfo.tiny, max=1.0 - finfo.eps)
```
